gateway-service/src/main/java/org/zowe/apiml/gateway/filters/PageRedirectionFilterFactory.java

```diff
@@ -48,65 +48,138 @@ public class PageRedirectionFilterFactory extends AbstractGatewayFilterFactory<P
+        if (locationUri.getHost() == null) {
+            return instance;
         }
 
-        return ((Stream<?>) discoveryClient.getInstances(config.serviceId).stream())
-            .findAny()
-                .map(ServiceInstance.class::cast)
-                .map(serviceInstance -> {
-                        Map<String, String> metadata = serviceInstance.getMetadata();
-                        RoutedServices routes = metadataParser.parseRoutes(metadata);
-                        try {
-                            String newUrl = transformService.transformURL(ServiceType.ALL, StringUtils.toRootLowerCase(config.serviceId), location, routes, false);
-                            if (isAttlsEnabled) {
-                                newUrl = UriComponentsBuilder.fromUriString(newUrl).scheme("https").build().toUriString();
-                            }
-                            return newUrl;
-                        } catch (URLTransformationException e) {
-                            log.debug("The URL for the redirect {} cannot be transformed: {}", location, e.getMessage());
-                            return "";
-                        }
-                    }
-                )
-                .orElse("");
+        URI hostUri = getHostUri(locationUri);
+        if (instance.map(i -> getHostUri(i).equals(hostUri)).orElse(false)) {
+            return instance;
+        }
+
+        return discoveryClient.getServices().stream()
+            .map(discoveryClient::getInstances)
+            .flatMap(List::stream)
+            .filter(i -> hostUri.equals(getHostUri(i)))
+            .findFirst();
     }
 
-    @Override
-    public GatewayFilter apply(Config config) {
-        return (exchange, chain) -> chain.filter(exchange)
-            .then(processNewLocationUrl(exchange, config));
+    private String normalizePath(String path) {
+        if (!path.startsWith("/")) {
+            path = "/" + path;
+        }
+        if (!path.endsWith("/")) {
+            path = path + "/";
+        }
+        return path;
     }
 
-    private Mono<Void> processNewLocationUrl(ServerWebExchange exchange, Config config) {
-        return Mono.fromCallable(() -> {
-            var response = exchange.getResponse();
-            if (response.getStatusCode().is3xxRedirection()) {
-                return getNewLocationUrl(config, response.getHeaders().getFirst(HttpHeaders.LOCATION));
+    private boolean isMatching(RoutedService route, URI uri) {
+        var servicePath = normalizePath(route.getServiceUrl());
+        var locationPath = normalizePath(uri.getPath());
+        return locationPath.startsWith(servicePath);
+    }
+
+    private Mono<Void> processNewLocationUrl(ServerWebExchange exchange, Config config, Optional<ServiceInstance> instance) {
+        var response = exchange.getResponse();
+        if (!response.getStatusCode().is3xxRedirection()) {
```

Comment:
```suggestion
        if (!Objects.requireNonNull(response.getStatusCode()).is3xxRedirection()) {
```

Reply:
I do not see a difference. I would throw a NullPointerException as before. I have rewritten with Optional.